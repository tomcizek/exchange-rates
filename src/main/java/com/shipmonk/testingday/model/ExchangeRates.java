package com.shipmonk.testingday.model;

public class ExchangeRates {
}
