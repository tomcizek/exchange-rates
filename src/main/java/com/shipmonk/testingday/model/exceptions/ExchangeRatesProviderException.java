package com.shipmonk.testingday.model.exceptions;

public class ExchangeRatesProviderException extends Exception {


    public ExchangeRatesProviderException(
        String message
    ) {
        super(message);
    }

    public ExchangeRatesProviderException(
        String message,
        Throwable cause
    ) {
        super(message, cause);
    }

    public static ExchangeRatesProviderException becauseNoConversionSymbolsProvided() {
        return new ExchangeRatesProviderException(
            "No symbols for currency conversion provided."
        );
    }

    public static ExchangeRatesProviderException becauseFetchingFailed(Exception e) {
        return new ExchangeRatesProviderException(
            "Fetching exchange rates failed.",
            e
        );
    }

    public static ExchangeRatesProviderException becauseParsingResultJSONFailed(Exception e) {
        return new ExchangeRatesProviderException(
            "Parsing JSON result of exchange rates failed.",
            e
        );
    }
}
