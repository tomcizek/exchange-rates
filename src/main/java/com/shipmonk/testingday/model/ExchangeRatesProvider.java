package com.shipmonk.testingday.model;

import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.time.LocalDate;
import java.util.List;
import javax.money.CurrencyUnit;

public interface ExchangeRatesProvider {

    ExchangeRates getRates(LocalDate day, List<CurrencyUnit> symbols) throws ExchangeRatesProviderException;

}
