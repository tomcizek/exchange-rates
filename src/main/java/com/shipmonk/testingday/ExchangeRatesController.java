package com.shipmonk.testingday;

import com.shipmonk.testingday.model.ExchangeRates;
import com.shipmonk.testingday.model.ExchangeRatesProvider;
import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.Arrays;
import java.util.List;
import javax.money.CurrencyUnit;
import javax.money.Monetary;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(
    path = "/api/v1/rates"
)
public class ExchangeRatesController {

    private ExchangeRatesProvider exchangeRatesProvider;

    @Value("${exchange.rates.currency.target.defaults}")
    private String defaultTargetCurrencies;

    public ExchangeRatesController(ExchangeRatesProvider exchangeRatesProvider) {
        this.exchangeRatesProvider = exchangeRatesProvider;
    }

    @RequestMapping(method = RequestMethod.GET, path = "/{day}")
    public ResponseEntity<Object> getRates(@PathVariable("day") String day) {

        LocalDate localDate;
        try {
            localDate = LocalDate.parse(day);
        } catch (DateTimeParseException e) {
            return new ResponseEntity<>(
                String.format("Invalid date format: '%s'", day),
                HttpStatus.BAD_REQUEST
            );
        }

        List<CurrencyUnit> defaultTargetCurrencies;
        try {
            defaultTargetCurrencies = Arrays.stream(this.defaultTargetCurrencies.split(","))
                .map(Monetary::getCurrency)
                .toList();
        } catch (Exception e) {
            // TODO log exception, bad configuration
            return new ResponseEntity<>(
                "Invalid currency symbols", HttpStatus.INTERNAL_SERVER_ERROR);
        }

        try {
            ExchangeRates rates = exchangeRatesProvider.getRates(
                localDate, defaultTargetCurrencies);
            return new ResponseEntity<>(rates, HttpStatus.OK);
        } catch (ExchangeRatesProviderException e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

}
