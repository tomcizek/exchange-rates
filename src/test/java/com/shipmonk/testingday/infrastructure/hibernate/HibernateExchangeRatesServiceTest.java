package com.shipmonk.testingday.infrastructure.hibernate;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.shipmonk.testingday.ApprovalTest;
import com.shipmonk.testingday.model.ExchangeRates;
import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.io.IOException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import javax.money.Monetary;
import javax.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

@SpringBootTest
@Transactional
class HibernateExchangeRatesServiceTest extends ApprovalTest {

    @Autowired
    private HibernateExchangeRatesService hibernateExchangeRatesService;


    @Test
    public void itShouldFailWhenNoConversionSymbolsProvided() {
        Executable getRatesCall = () -> hibernateExchangeRatesService.getRates(
            LocalDate.of(2023, 6, 6),
            List.of()
        );

        var exception = assertThrows(
            ExchangeRatesProviderException.class,
            getRatesCall
        );

        assertEquals("No symbols for currency conversion provided.", exception.getMessage());
    }

    @Test
    public void itShouldReturnEmptyListOfExchangeRatesWhenNoneFoundInDb()
        throws ExchangeRatesProviderException, IOException, InterruptedException {

        var exchangeRates = hibernateExchangeRatesService.getRates(
            LocalDate.of(2023, 6, 6),
            List.of(
                Monetary.getCurrency("EUR"),
                Monetary.getCurrency("CZK")
            )
        );

        this.thenExpectedResult(exchangeRates);
    }

    @Test
    public void itShouldReturnExchangeRatesPresentInDatabase()
        throws ExchangeRatesProviderException, IOException, InterruptedException {

        var exchangeRatesToInsert = new ExchangeRates(
            true,
            "EUR",
            LocalDate.of(2023, 6, 6),
            Map.of("CZK", new BigDecimal("23.578329"))
        );
        hibernateExchangeRatesService.saveRates(exchangeRatesToInsert);

        var exchangeRates = hibernateExchangeRatesService.getRates(
            LocalDate.of(2023, 6, 6),
            List.of(
                Monetary.getCurrency("EUR"),
                Monetary.getCurrency("CZK")
            )
        );

        this.thenExpectedResult(exchangeRates);
    }

    @Test
    public void itShouldReturnMultipleExchangeRatesPresentInDatabase()
        throws ExchangeRatesProviderException, IOException, InterruptedException {

        var exchangeRatesToInsert = new ExchangeRates(
            true,
            "EUR",
            LocalDate.of(2023, 6, 6),
            Map.of(
                "CZK", new BigDecimal("23.578329"),
                "JPY", new BigDecimal("114.44321")
            )
        );
        hibernateExchangeRatesService.saveRates(exchangeRatesToInsert);

        var exchangeRates = hibernateExchangeRatesService.getRates(
            LocalDate.of(2023, 6, 6),
            List.of(
                Monetary.getCurrency("CZK"),
                Monetary.getCurrency("JPY")
            )
        );

        this.thenExpectedResult(exchangeRates);
    }

}
