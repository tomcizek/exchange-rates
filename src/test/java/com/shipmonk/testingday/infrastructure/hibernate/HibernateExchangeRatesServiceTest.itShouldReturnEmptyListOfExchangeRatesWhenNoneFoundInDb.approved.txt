{
  "success" : true,
  "base" : "EUR",
  "date" : "2023-06-06",
  "rates" : {
    "CZK" : 23.58
  }
}
